### ISSUE TITLE
[Problem]: multiroom audio is sometimes a few milliseconds out of sync

### ISSUE THREAD

**Issue by dasl-** [mikebrady/shairport-sync#1563]
### What happened?

Hi again! I've been using the multiroom feature of SPS more lately. Specifically, I'm connecting to two SPS receivers in adjacent rooms. The receivers' names are `piwall10` and `kitchen`. When I connect to these two receivers from my iPhone (iOS 15.6.1), playing music via the Spotify iOS app, the sound is sometimes a few milliseconds out of sync. To my ears, it sounds like some amount of synchronization delay between 100ms and 1000ms if I had to guess. I am usually able to fix the problem by pressing the Pause button and then resuming playback immediately. 

When resuming playback, if it resumes in sync, it appears to stay in sync indefinitely. Likewise, if it resumes out of sync, it appears to stay out of sync indefinitely. Pressing the Pause button and then resuming playback immediately in a loop suggests we get synchronization issues something like 33% of the time when playback is resumed.

Interestingly, I am unable to reproduce this issue when playing local music on my iPhone's Music app. This might suggest the problem is something to do with realtime audio vs buffered audio (I believe spotify will use realtime audio and the Music app will use buffered audio).

## logs
See here for two log files: one for the `piwall10` receiver, and one for the `kitchen` receiver: https://gist.github.com/dasl-/9a6491f88e88644d5124b848714f0077

These logs show everything that happened since the SPS service was restarted on both raspberry pis. I started playing the song Thriller by Michael Jackson via iOS spotify on both receivers. It started in sync. Then I pressed Pause and resumed playback a couple of times until it resumed playback out of sync. I believe it resumed playback out of sync at the timestamp `Oct 27 00:51:22` in these log files.

## audio recording demonstrating out of sync
See this audio recording of the out of sync audio: https://drive.google.com/file/d/1YzzzNEiU0xHgkzJDgybrfofE2UcJVtzg/view?usp=sharing

This recording was taken at the same time as the out of sync issue from the logs above. You'll notice the out of sync audio especially on the electronic snare drum. Here is how the song is supposed to sound: https://open.spotify.com/track/2LlQb7Uoj1kKyGhlkBf9aC?si=fb6959c53a6f462b

### Relevant log output

```shell
See above.
```

### Configuration Information.

piwall10 receiver information:
```
[piwall10] ~ % shairport-sync --displayConfig
         0.001108871 "shairport.c:1420" default metadata_pipename is "/tmp/shairport-sync-metadata".
         0.000249740 "shairport.c:2057" >> Display Config Start.
         0.008353574 "shairport.c:2057"
         0.000061445 "shairport.c:2057" From "uname -a":
         0.000019259 "shairport.c:2057"  Linux piwall10 5.10.103-v7l+ #1529 SMP Tue Mar 8 12:24:00 GMT 2022 armv7l GNU/Linux
         0.010979611 "shairport.c:2057"
         0.000068482 "shairport.c:2057" From /etc/os-release:
         0.000019870 "shairport.c:2057"  Raspbian GNU/Linux 10 (buster)
         0.009235537 "shairport.c:2057"
         0.000062037 "shairport.c:2057" From /sys/firmware/devicetree/base/model:
         0.000019926 "shairport.c:2057"  Raspberry Pi 4 Model B Rev 1.2
         0.000067463 "shairport.c:2057"
         0.000016833 "shairport.c:2057" Shairport Sync Version String:
         0.000016167 "shairport.c:2057"  4.1-rc0-159-g70efd5fc-AirPlay2-OpenSSL-Avahi-ALSA-soxr-metadata-dbus-sysconfdir:/etc
         0.000018315 "shairport.c:2057"
         0.000015241 "shairport.c:2057" Command Line:
         0.000015444 "shairport.c:2057"  shairport-sync --displayConfig
         0.000071333 "shairport.c:2057"
         0.000018408 "shairport.c:2057" Configuration File:
         0.000014555 "shairport.c:2057"  /etc/shairport-sync.conf
         0.000015630 "shairport.c:2057"
         0.000146167 "shairport.c:2057" Configuration File Settings:
         0.000021500 "shairport.c:2057"  general :
         0.000016018 "shairport.c:2057"  {
         0.000015871 "shairport.c:2057"    volume_range_db = 40;
         0.000016111 "shairport.c:2057"    volume_max_db = 0.0;
         0.000016222 "shairport.c:2057"    name = "piwall";
         0.000015685 "shairport.c:2057"  };
         0.000014222 "shairport.c:2057"  alsa :
         0.000015037 "shairport.c:2057"  {
         0.000015667 "shairport.c:2057"    output_device = "hw:Headphones";
         0.000014648 "shairport.c:2057"    mixer_control_name = "Headphone";
         0.000015537 "shairport.c:2057"  };
         0.000015759 "shairport.c:2057"  diagnostics :
         0.000014019 "shairport.c:2057"  {
         0.000015111 "shairport.c:2057"    log_verbosity = 2;
         0.000016074 "shairport.c:2057"    statistics = "yes";
         0.000014074 "shairport.c:2057"  };
         0.000015111 "shairport.c:2057"
         0.000015204 "shairport.c:2057" >> Display Config End.
         0.000015982 "shairport.c:2059" >> Goodbye!
         0.000028518 "shairport.c:1534" exit function called...
         0.000016796 "shairport.c:1552" Stopping D-Bus service
         0.000037463 "dbus-service.c:1131" stopping dbus service
         0.000063074 "shairport.c:1554" Stopping D-Bus service done
         0.000014871 "shairport.c:1570" Stopping DACP Monitor
         0.000014259 "shairport.c:1572" Stopping DACP Monitor Done
         0.000015241 "shairport.c:1576" Stopping metadata hub
         0.000016611 "shairport.c:1578" Stopping metadata done
         0.000015407 "shairport.c:1582" Stopping metadata
         0.000032760 "shairport.c:1584" Stopping metadata done
         0.000018907 "shairport.c:1586" Stopping the activity monitor.
         0.000015889 "shairport.c:1588" Stopping the activity monitor done.
         0.000031315 "shairport.c:1670" normal exit

[piwall10] ~ % nqptp -V
Version: 1.1-dev-170-gc71b49a. Shared Memory Interface Version: 8.
```

kitchen receiver information
```
[kitchen] ~ % shairport-sync --displayConfig
         0.000771981 "shairport.c:1420" default metadata_pipename is "/tmp/shairport-sync-metadata".
         0.000145445 "shairport.c:2057" >> Display Config Start.
         0.005040759 "shairport.c:2057"
         0.000045037 "shairport.c:2057" From "uname -a":
         0.000018834 "shairport.c:2057"  Linux kitchen 5.15.61-v7l+ #1579 SMP Fri Aug 26 11:13:03 BST 2022 armv7l GNU/Linux
         0.010871037 "shairport.c:2057"
         0.000053592 "shairport.c:2057" From /etc/os-release:
         0.000024537 "shairport.c:2057"  Raspbian GNU/Linux 11 (bullseye)
         0.007809352 "shairport.c:2057"
         0.000041907 "shairport.c:2057" From /sys/firmware/devicetree/base/model:
         0.000018537 "shairport.c:2057"  Raspberry Pi 4 Model B Rev 1.2
         0.000070982 "shairport.c:2057"
         0.000019315 "shairport.c:2057" Shairport Sync Version String:
         0.000016778 "shairport.c:2057"  4.1-rc0-159-g70efd5fc-AirPlay2-OpenSSL-Avahi-ALSA-soxr-metadata-dbus-sysconfdir:/etc
         0.000019851 "shairport.c:2057"
         0.000015815 "shairport.c:2057" Command Line:
         0.000016000 "shairport.c:2057"  shairport-sync --displayConfig
         0.000047037 "shairport.c:2057"
         0.000018204 "shairport.c:2057" Configuration File:
         0.000016537 "shairport.c:2057"  /etc/shairport-sync.conf
         0.000016463 "shairport.c:2057"
         0.000143889 "shairport.c:2057" Configuration File Settings:
         0.000020704 "shairport.c:2057"  general :
         0.000016685 "shairport.c:2057"  {
         0.000016240 "shairport.c:2057"    volume_range_db = 40;
         0.000016482 "shairport.c:2057"    volume_max_db = 0.0;
         0.000016685 "shairport.c:2057"    name = "pi %h";
         0.000016222 "shairport.c:2057"  };
         0.000015963 "shairport.c:2057"  alsa :
         0.000015741 "shairport.c:2057"  {
         0.000016204 "shairport.c:2057"    output_device = "hw:Headphones";
         0.000016444 "shairport.c:2057"    mixer_control_name = "Headphone";
         0.000016148 "shairport.c:2057"  };
         0.000016556 "shairport.c:2057"  diagnostics :
         0.000015759 "shairport.c:2057"  {
         0.000017259 "shairport.c:2057"    log_verbosity = 2;
         0.000016371 "shairport.c:2057"    statistics = "yes";
         0.000016129 "shairport.c:2057"  };
         0.000016000 "shairport.c:2057"
         0.000015426 "shairport.c:2057" >> Display Config End.
         0.000016463 "shairport.c:2059" >> Goodbye!
         0.000023130 "shairport.c:1534" exit function called...
         0.000017000 "shairport.c:1552" Stopping D-Bus service
         0.000043074 "dbus-service.c:1131" stopping dbus service
         0.000022574 "shairport.c:1554" Stopping D-Bus service done
         0.000017389 "shairport.c:1570" Stopping DACP Monitor
         0.000016611 "shairport.c:1572" Stopping DACP Monitor Done
         0.000016111 "shairport.c:1576" Stopping metadata hub
         0.000016000 "shairport.c:1578" Stopping metadata done
         0.000014445 "shairport.c:1582" Stopping metadata
         0.000014740 "shairport.c:1584" Stopping metadata done
         0.000014074 "shairport.c:1586" Stopping the activity monitor.
         0.000016389 "shairport.c:1588" Stopping the activity monitor done.
         0.000027019 "shairport.c:1670" normal exit

[kitchen] ~ % nqptp -V
Version: 1.1-dev-170-gc71b49a. Shared Memory Interface Version: 8.
```

### How did you install Shairport Sync?

Built from source

### Check previous issues

- [X] Confirm

**Comment by mikebrady** [comment]
Thanks for this report. Not getting this on my setups, I'm afraid, which include Pis running on both of those OS versions, but my iOS devices are fully up to date. Can you say what kind of amplifiers your Pis are connected in to please?

Given the problems that have surfaced with Spotify recently, e.g. #1551, it would be easy to be suspicious of the Spotify app. 

To check this, there are other apps that also use Realtime streams and it would be worth seeing if the problem occurs with them too. The app I use is the "Radio Swiss Classic" app by SRG SSR (not the similarly-titled "Radio Swiss Classic App" which is by a third party).

**Comment by dasl-** [comment]
I am interested in upgrading to iOS 16 to see if it solves this issue, but I noticed that you mentioned there are is an airplay bug with iOS 16, regarding the 2,112 extra frames: https://github.com/mikebrady/shairport-sync/discussions/1332

I didn't quite understand how severe / impactful this bug is - it sounds like there might by playback glitches for the first 30 seconds of play until SPS can recover? Would that happen every time I unpause audio? I'm not sure how easy it is to reverse an iOS upgrade, so I'm not eager to upgrade if it will cause even more problems 😅

**Comment by mikebrady** [comment]
> I am interested in upgrading to iOS 16 to see if it solves this issue, but I noticed that you mentioned there are is an airplay bug with iOS 16, regarding the 2,112 extra frames: https://github.com/mikebrady/shairport-sync/discussions/1332

Thanks for the reminder -- I just added a quick note on the What's New discussion.

Happily, it _seems_ to be gone on iOS 16.1, iPadOS 16.1  and Ventura. Sometimes there are occasional glitches on Ventura, but I can't see any pattern yet except that they only seem to occur when streaming, and not if you're playing something that is already downloaded.

**Comment by mikebrady** [comment]
Thanks for all the information about speakers. It does seem to suggest that there is something about Realtime streams alright. Are you connecting via the Pi's audio jack or via the HDMI port? (Just FYI, my systems only use the built-in audio for testing -- for real listening they use I2S devices like the Pimoroni PHAT DAC or the IQaudio DigiAMP+ or else USB.)

So, can you give me a step-by-step of your torture tests please? E.g.:
1. On the iPhone, select both speakers in the AirPlay panel of the Control Center.
2. Play something from Spotify.
3. Pause Spotify.
4. Resume Spotify
5. etc...
6.

Or, do you deselect and reselect a speaker each time? Or, do you start Spotify first and then select a speaker and then another... , or... 🙂. The sequence might be important...

**Comment by mikebrady** [comment]
Crazy thought -- could it be that WiFi receivers are running in power-saving mode? (See [here](https://github.com/mikebrady/shairport-sync/blob/master/TROUBLESHOOTING.md#wifi-adapter-running-in-power-saving--low-power-mode).)

**Comment by dasl-** [comment]
>could it be that WiFi receivers are running in power-saving mode? 

I don't think so. I [disable power save when I install SPS](https://github.com/dasl-/pitools/blob/334f76fecf0fea785587b8df834032e6fa351435/shairport-sync/install_or_update_shairport_sync.sh#L93). 
```
[kitchen] ~ % sudo iw dev wlan0 get power_save
Power save: off
```

Thanks for the info about iOS 16.1! I've just upgraded to iOS 16.1. I am still able to reproduce this issue on iOS 16.1 unfortunately.

>Are you connecting via the Pi's audio jack or via the HDMI port?

I use the pi's built in audio jack on all my pis.

>So, can you give me a step-by-step of your torture tests please? E.g.:

Yes definitely. I've recorded a video here of me reproducing the bug on my iphone running iOS 16.1: https://drive.google.com/file/d/1if6IX4siC3IyQuYK6HEPaWvJMpjKoBWb/view

TLDW:
1) open the swiss radio classic app
2) start playing radio, initially through the iphone builtin speakers
3) press the button on the side of the iphone to turn off the screen
4) tap iphone to show lock screen
5) connect to SPS receiver 1 via airplay lock screen menu
6) connect to SPS receiver 2 via airplay lock screen menu
7) pause via lockscreen button
8) unpause via lockscreen button
9) repeat (7) and (8) as necessary to trigger the bug

The bug is more easily detectable to my ears when playing music with drums, rather than classical string music. So I find it easier to detect when playing music like that on spotify, rather than swiss radio classic.

Also, although I included specific steps here, in my experience the bug is robustly reproducible no matter how I connect to airplay, and no matter whether I do things via the lock screen, or in the app itself. E.g. if I connect to airplay via spotify's in-app menu rather than the lock screen, and if I pause/unpause in the app itself versus on the lock screen, the bug is reproducible.

I realize now that in the video, when the first big sync issue occured, I was paused while radio transitioned to the next track. I don't think this is important to the reproduction of the bug. I have reproduced this other times in the middle of a given track.

**Comment by jdtsmith** [comment]
> One is a standalone surround sound amplifier 

Most surround processors introduce an additional lag as they process/mix/redirect two channel input for "immersive" output, even using an APE as the AP2 decoder.  On my Denon, there is a stereo "direct sound" mode which cuts out this delay.  It's a small bummer because it means you can't pull off the low-frequency info and send to the subs, but at least it's all in sync again.

**Comment by dasl-** [comment]
Hello again, I've been doing some more testing on this issue. Some minor updates:

1. I tested with the Radio Swiss Pop iOS app (by SRG SSR) to further confirm that this synchronization issue is not limited to the spotify app. In my opinion, the Radio Swiss Pop app plays songs in which it is much easier to detect synchronization issues than the Radio Swiss Classic app (i.e. songs with drums and vocals)
2. I've performed even more testing with buffered vs realtime streams which further confirms that this synchronization is limited to realtime streams
3. I've performed many tests in which both SPS receivers are connected via ethernet to rule out wifi related issues on the server side, although theoretically there could still be wifi issues on the client side - the client is an iphone connected via wifi.
4. I've performed test both with and without setting `audio_backend_buffer_desired_length_in_seconds = 2`. Interestingly, it is much easier to find these synchronization issues when `audio_backend_buffer_desired_length_in_seconds = 2` is omitted from my configuration. When it is omitted from my configuration, I estimate that I am able to find synchronization issues approximately 50% of the time when I do a "pause/resume" loop. And furthermore, when these synchronization issues occur, they are on the order of hundreds of milliseconds out of sync, by my estimate. However, when I have `audio_backend_buffer_desired_length_in_seconds = 2` in my configuration on both SPS receivers, these synchronization issues either become much rarer and/or they become much smaller in size, i.e. when they occur, they are on the order of less than 100 ms out of sync by my estimate.

It's hard for me to know what to make of all this. Point 1 seems to suggest it is not related to a particular app. Point 2 suggests that it is not related to any lag related to my audio amplifiers. I'm not exactly sure what to make of Point 4.

I'm curious if you are still unable to reproduce? If you are unable to reproduce, maybe this suggest a problem with either:

1. the wifi connection from my iphone to the router
2. the built-in raspberry pi soundcard (since you use a different soundcard)

**Comment by dasl-** [comment]
Hi again, I'm back with another update. I added some logging to see if I could get more information about these multiroom realtime audio synchronization issues. Here are my logging modifications: https://gist.github.com/dasl-/bbb9d2806c80e065c50063b30329412f

I've been doing some research into the code, and my understanding is that we should expect sound that has been sent to the DAC via `alsa_pcm_write` to be audible after a variable delay. The delay (in number of frames) can be obtained by calling `delay_and_status`. Please feel free to correct my understanding if I'm mistaken.

I obtained logs from 2 SPS receivers (`piwall` and `kitchen`) during a period when my iphone client was connected to them playing spotify. During the period that these logs cover, I observed that the playback was out of sync. Specifically, I estimate that the `kitchen` receiver was playing about 250 milliseconds ahead of the `piwall` receiver.

* [`piwall` logs](https://gist.githubusercontent.com/dasl-/d422db1decaf6d397f76b2f941adfc65/raw/fae5fb0c8fd4d23898b12ebe7ebbd55ef6bef657/piwall.logs)
* [`kitchen` logs](https://gist.githubusercontent.com/dasl-/61216ccb8e873867fd699659d8e8b672/raw/dbfd4bbffbcb238a4ee2ade4b5f7d608b76818a1/kitchen.log)

From these logs, it seems to me that both SPS receivers were calling `alsa_pcm_write` at the correct times for each audio buffer, i.e. the `deviation from requested playtime` in my logs was under 1 millisecond.  This suggests to me that the problem is not network related. 

At this point, I suspect the problem could be related to one of two things:

1) Perhaps something is wrong with the timing code - maybe [`playtime`](https://github.com/mikebrady/shairport-sync/blob/99e76fdf2daad00aeaba065738825c7a82d573fd/audio_dummy.c#L51) is incorrect somehow.
2) Perhaps something is wrong with alsa. Maybe the delay reported by `delay_and_status` is incorrect somehow.

Given that you do not seem to encounter this issue with your setup, which as I understand it, you do not use the raspberry pi built-in soundcard, I am leaning toward suspecting (2) as the cause. Very curious if you have any thoughts on all of this.

**Comment by dasl-** [comment]
I believe I have solved my problem. 

**TLDR:** adding `disable_standby_mode = "always";` to the `alsa` stanza of the shairport-sync config file seems to fix the problem.

Tonight I decided to give this very annoying issue one more debugging session. I was browsing the `audio_alsa.c` code, and on a whim I decided try preventing the [`sub_flush`](https://github.com/mikebrady/shairport-sync/blob/a3151c3c55b6e2b61906bc57c1ed0d38b2cb17c7/audio_alsa.c#L1957) code from running. I realized I could get this behavior by adding `disable_standby_mode = "always";` to the `alsa` stanza of my config file. So I modified my config file, restarted shairport-sync, and now I can no longer reproduce this issue.

I don't have a good sense of why this solves the problem, so if anyone has any theories, I'd be interested to hear. Furthermore, I wonder if the default value of `disable_standby_mode` should be `always`. A quick search of github issues showed that at least 2 issues seemed to be solved by setting `disable_standby_mode = "always"`:

1. https://github.com/mikebrady/shairport-sync/issues/1363#issuecomment-1126879833
2. https://github.com/mikebrady/shairport-sync/issues/1427#issuecomment-1040250153

I guess the [sample config comments](https://github.com/mikebrady/shairport-sync/blob/54cb9de1836fcb764bbae7397b6f79899f674268/scripts/shairport-sync.conf#L114) do have some comments on the default value: `Default is "never", but only for backwards compatibility`. I am not sure what the downsides / tradeoff of breaking this backward compatibility would be.

**Comment by mrkwagga** [comment]
> I believe I have solved my problem.
> 
> **TLDR:** adding `disable_standby_mode = "always";` to the `alsa` stanza of the shairport-sync config file seems to fix the problem.
> 
> Tonight I decided to give this very annoying issue one more debugging session. I was browsing the `audio_alsa.c` code, and on a whim I decided try preventing the [`sub_flush`](https://github.com/mikebrady/shairport-sync/blob/a3151c3c55b6e2b61906bc57c1ed0d38b2cb17c7/audio_alsa.c#L1957) code from running. I realized I could get this behavior by adding `disable_standby_mode = "always";` to the `alsa` stanza of my config file. So I modified my config file, restarted shairport-sync, and now I can no longer reproduce this issue.
> 
> I don't have a good sense of why this solves the problem, so if anyone has any theories, I'd be interested to hear. Furthermore, I wonder if the default value of `disable_standby_mode` should be `always`. A quick search of github issues showed that at least 2 other issues seemed to be solved by setting `disable_standby_mode = "always"`:
> 
>     1. [Shairport-sync Development crashes now and then #1363 (comment)](https://github.com/mikebrady/shairport-sync/issues/1363#issuecomment-1126879833)
> 
>     2. [No sound (crash?) after scrubbing through song #1427 (comment)](https://github.com/mikebrady/shairport-sync/issues/1427#issuecomment-1040250153)
> 
> 
> I guess the [sample config comments](https://github.com/mikebrady/shairport-sync/blob/54cb9de1836fcb764bbae7397b6f79899f674268/scripts/shairport-sync.conf#L114) do have some comments on the default value: `Default is "never", but only for backwards compatibility`. I am not sure what the downsides / tradeoff of breaking this backward compatibility would be.

I'm eager to try this, I'll test and provide feedback.

**Comment by mikebrady** [comment]
Thanks @dasl-, as ever.

I meant to ask -- what kind of network are you running? Specifically, is it a "traditional" network with a single WiFi access point and, maybe, with some Ethernet ports on the router, or is it any kind of mesh network?

**Comment by dasl-** [comment]
Hey Mike, here is a diagram of my network setup:

<img width="932" alt="Screen Shot 2022-11-28 at 11 37 57 AM" src="https://user-images.githubusercontent.com/1332576/204332224-653e7d2d-e2c2-4101-836f-0a1c85f2f52f.png">

**Comment by dasl-** [comment]
@mrkwagga I'm curious if you got a chance to try my fix yourself yet?

Also, I'm going to close this issue, because I have solved the problem on my setup (see https://github.com/mikebrady/shairport-sync/issues/1563#issuecomment-1328166125 )

**Comment by bai-yi-bai** [comment]
Hello, I can report that I am having the same issue on an RPI3B. Playback on my RPI4 and AppleTV 4K are in sync, but the RPI3B is a few milliseconds behind. 

@dasl- 's fix did not work for me.

Both the RPI4 and RPI3B have fresh installations of Raspberry Pi OS Lite (64-bit) (Debian Bullseye). I used the Raspberry Pi Imager v1.7.3 to write the SD cards and configure their interfaces.
Initially, I used Wi-Fi to set up both devices via ssh, but later switched to Ethernet (same 192.168.x.x subnet, although both are also picking up IPv6 addresses also). The software configurations are almost identical, except I disabled the RPI3B's LEDs. 

Attempted Fix:
* I disabled/stopped the systemd service.
* I modified the shairport-sync.conf file (un-commented out `disable_standby_mode = "never";` and changed it to `disable_standby_mode = "always";`
* I ran `shairport-sync -c shairport-sync.conf`, but the delay was still present.

The only difference I see between the RPI3 and RPI4 is that the hardware is different, so I'm betting something "in the guts" of Raspberry Pi OS/Bullseye has been changed for RPI3B.

I am going to try setting up some different OSes for the RPI3, perhaps and older version of Raspberry Pi OS Lite (64-bit), Raspberry Pi OS Lite (32-bit), Ubuntu Server 22.10 (both 32-bit and 64-bit), Apertis, LibELEC, Volumio, or OSMC.

**Comment by dasl-** [comment]
@bai-yi-bai that's strange - for me the `disable_standby_mode = "always";` fix is working both for raspberry pi 4b, 3b, and 3a.

```
% cat /proc/cpuinfo | grep Model
Model		: Raspberry Pi 3 Model B Plus Rev 1.3

% cat /proc/cpuinfo | grep Model
Model		: Raspberry Pi 3 Model A Plus Rev 1.0

% cat /proc/cpuinfo | grep Model
Model		: Raspberry Pi 4 Model B Rev 1.2
```

I am running raspberry pi OS Lite 32-bit on all my pis though - not 64 bit.

```
% uname -a
Linux study 5.15.76-v7+ #1597 SMP Fri Nov 4 12:13:17 GMT 2022 armv7l GNU/Linux
% getconf LONG_BIT
32
```

Regarding your attempted fix, it might rule out some simple mistake if you can show us the output of `shairport-sync --displayConfig`. Here is mine, for reference: https://gist.github.com/dasl-/024f2284681c818a12181ca895bbbb41

**Comment by bai-yi-bai** [comment]
Thank you for the quick reply. 

I notice a few differences from yours:

- `aarch64 GNU/Linux` vs. `armv7l (expected)`
- `Raspberry Pi 3 Model B Rev 1.2`
- `4.1.1-2-g6240fd83-AirPlay2-OpenSSL-Avahi-ALSA-soxr-sysconfdir:/etc` vs. `4.1-rc0-273-gf164f19a-AirPlay2-OpenSSL-Avahi-ALSA-soxr-metadata-dbus-sysconfdir:/etc`
- I took your `volume_range_db = 40;` setting after reading in some threads that the volume range on RPIs is a little optimistic about what the DAC can actually produce.

I compiled from the current files on git; I read that "metadata" is recommended for daapd/owntone. I plan on seeing if I can set up pipes.

```pi@blueberrypi:~/shairport-sync $ uname -a
Linux blueberrypi 5.15.84-v8+ #1613 SMP PREEMPT Thu Jan 5 12:03:08 GMT 2023 aarch64 GNU/Linux
pi@blueberrypi:~/shairport-sync $ shairport-sync --displayConfig
>> Display Config Start.

From "uname -a":
 Linux blueberrypi 5.15.84-v8+ #1613 SMP PREEMPT Thu Jan 5 12:03:08 GMT 2023 aarch64 GNU/Linux

From /etc/os-release:
 Debian GNU/Linux 11 (bullseye)

From /sys/firmware/devicetree/base/model:
 Raspberry Pi 3 Model B Rev 1.2

Shairport Sync Version String:
 4.1.1-2-g6240fd83-AirPlay2-OpenSSL-Avahi-ALSA-soxr-sysconfdir:/etc

Command Line:
 shairport-sync --displayConfig

Configuration File:
 /etc/shairport-sync.conf

Configuration File Settings:
 general :
 {
   volume_range_db = 40;
 };
 alsa :
 {
   disable_standby_mode = "always";
 };

>> Display Config End.
>> Goodbye!
```

**Comment by bai-yi-bai** [comment]
Okay, looks like installing the 32-bit version worked. It seemed like the first couple seconds of playback were out of sync, but then I noticed some re-syncing in the console output. Thank you! Now to find cheap, low-power devices to spread throughout my home.

```
pi@lemonberrypi:~ $ sudo shairport-sync --displayConfig
>> Display Config Start.

From "uname -a":
 Linux lemonberrypi 5.15.61-v7+ #1579 SMP Fri Aug 26 11:10:59 BST 2022 armv7l GNU/Linux

From /etc/os-release:
 Raspbian GNU/Linux 11 (bullseye)

From /sys/firmware/devicetree/base/model:
 Raspberry Pi 3 Model B Rev 1.2

Shairport Sync Version String:
 4.1.1-2-g6240fd83-AirPlay2-OpenSSL-Avahi-ALSA-soxr-metadata-sysconfdir:/etc

Command Line:
 shairport-sync --displayConfig

Configuration File:
 /etc/shairport-sync.conf

Configuration File Settings:
 general :
 {
   volume_range_db = 40;
 };
 alsa :
 {
   disable_standby_mode = "always";
 };

>> Display Config End.
>> Goodbye!
```

```
7.911712028 "player.c:2591" Sync Error ms | Net Sync PPM | All Sync PPM | Missing |   Late | Too Late | Resend Reqs | Min DAC Queue | Min Buffers | Max Buffers | Received FPS | Output FPS (r) | Output FPS (c)
0.000639375 "player.c:2591"          0.25           -8.6            8.6         0        0          0             0            7424           216           232       44491.43              N/A              N/A
8.010734372 "player.c:2591"         -0.19          -14.2           14.2         0        0          0             0            7374           217           233       44222.93         44104.96         44104.99
8.010017810 "player.c:2591"         -0.15          -11.3           11.3         0        0          0             0            7378           217           233       44169.48         44102.48         44102.50
8.016745101 "player.c:2591"          0.07          -31.2           31.2         0        1          0             1            7402           215           233       44145.98         44101.71         44101.74
7.987299893 "player.c:2591"         -1.37          -11.3           11.3         0        1          0             1            7341           217           232       44148.35         44101.32         44101.35
8.007262444 "player.c:2591"         -0.52          -31.2           31.2         0        2          0             2            7383           217           233       44147.27         44101.10         44101.13
8.019392237 "player.c:2591"          1.02         -422.0          422.0         0        2          0             2            7334           216           233       44147.36         44100.93         44100.96
7.989713018 "player.c:2591"          1.53          -48.2           48.2         0        2          0             2            7465           217           232       44128.25         44100.81         44100.84
8.009100830 "player.c:2591"          0.90          -25.5           25.5         0        2          0             2            7417           216           232       44135.34         44100.72         44100.75
8.000489320 "player.c:2591"          0.19          -14.2           14.2         0       18          0             9            7421           209           232       44121.31         44100.63         44100.66
8.015413382 "player.c:2591"         -0.18          -31.2           31.2         0       43          0            13            7400           216           232       44118.20         44100.60         44100.63
8.009717653 "player.c:2591"         -0.59          -36.8           36.8         0       47          0            17            7361           217           232       44122.14         44100.55         44100.58
8.007068643 "player.c:2591"         -1.16          -17.0           17.0         0       48          0            18            7359           217           232       44116.39         44100.53         44100.56
7.999607549 "player.c:2591"         -1.15           -8.5            8.5         0       49          0            19            7356           192           232       44120.96         44100.48         44100.51
8.004016612 "player.c:2591"         -1.59           -5.7           34.0         0       57          0            27            7329           193           233       44114.14         44100.46         44100.49
8.011633955 "player.c:2591"         -1.81          -11.3           22.7         0       57          0            27            7323           217           232       44111.20         44100.44         44100.47
8.011190987 "player.c:2591"         -1.85           14.2           31.2         0       57          0            27            7321           217           232       44111.38         44100.41         44100.44
7.988603851 "player.c:2591"         -1.93           28.3           51.0         0       68          0            38            7315           218           232       44111.50         44100.39         44100.42
8.003297549 "player.c:2591"         -1.81           -5.7           22.7         0       71          0            41            7327           216           232       44114.26         44100.38         44100.40
```


```
7.913831847 "player.c:2591" Sync Error ms | Net Sync PPM | All Sync PPM | Missing |   Late | Too Late | Resend Reqs | Min DAC Queue | Min Buffers | Max Buffers | Received FPS | Output FPS (r) | Output FPS (c)
0.000129648 "player.c:2591"          0.47            0.0            0.0         0        0          0             0            7426           218           233       44421.05              N/A              N/A
8.011054643 "player.c:2591"          0.07            0.0            0.0         0        0          0             0            7393           217           233       44242.20         44099.91         44100.60
8.009608255 "player.c:2591"          0.54            0.0            0.0         0        0          0             0            7418           218           233       44182.03         44099.92         44100.61
8.016145014 "player.c:2591"          1.23            0.0            0.0         0        0          0             0            7424           217           233       44144.73         44099.91         44100.59
8.009440217 "player.c:2591"         -1.35            0.0            0.0         0        0          0             0            7339           217           232       44140.30         44099.91         44100.59
7.983622366 "player.c:2591"          0.22            2.8            2.8         0        0          0             0            7399           218           234       44141.22         44099.91         44100.58
8.017340365 "player.c:2591"          1.66         -277.6          277.6         0        0          0             0            7470           217           233       44138.93         44099.88         44100.56
8.011352884 "player.c:2591"          1.35            0.0            0.0         0        0          0             0            7461           218           233       44082.79         44099.87         44100.54
8.005488496 "player.c:2591"          0.75            0.0            0.0         0        0          0             0            7434           217           233       44125.27         44099.89         44100.56
7.990706587 "player.c:2591"          0.22            0.0            0.0         0        0          0             0            7422           217           233       44122.37         44099.87         44100.54
8.009001699 "player.c:2591"         -0.01            0.0            0.0         0        0          0             0            7407           218           233       44122.67         44099.89         44100.56
8.012564273 "player.c:2591"         -0.12            0.0            0.0         0        0          0             0            7411           217           233       44121.29         44099.88         44100.55
8.010650162 "player.c:2591"         -0.23            0.0            0.0         0        0          0             0            7422           218           233       44113.99         44099.88         44100.54
7.995762440 "player.c:2591"         -0.60            0.0            0.0         0        0          0             0            7374           193           233       44117.34         44099.87         44100.53
8.002445810 "player.c:2591"         -1.15            0.0            0.0         0        0          0             0            7348           193           233       44117.25         44099.87         44100.53
8.011255032 "player.c:2591"         -1.41            0.0            0.0         0        0          0             0            7348           217           233       44114.18         44099.87         44100.52
8.012063440 "player.c:2591"         -1.41            0.0            0.0         0        0          0             0            7358           217           233       44113.84         44099.88         44100.53
8.005544662 "player.c:2591"         -1.65            0.0            0.0         0        0          0             0            7333           218           233       44113.69         44099.87         44100.53
8.020670884 "player.c:2591"         -1.92           17.0           17.0         0        0          0             0            7329           217           233       44111.09         44099.87         44100.53
```

**Comment by dasl-** [comment]
Nice! I was about to write that nothing stood out to me as obviously wrong in your configuration. Trying the 32-bit OS seemed like a good idea. Glad it is working now.

**Comment by bai-yi-bai** [comment]
Update: I did some more testing and noticed the RPI4 was now out of sync with the RPI3 and AppleTV, so I disabled standby on that platform, too.

**Comment by dasl-** [comment]
Good to know - thanks for confirming that `disable_standby_mode = "always";` is solving the out of sync multiroom playback issue for you too. I was starting to wonder if I was the only one with this problem 😅 .

Also thanks for determining that this fix only works on raspberry pi OS 32 bit, rather than 64 bit.

**Comment by bai-yi-bai** [comment]
Oh, let me clarify: my RPI4 is running 5.15.84-v8+ #1613 SMP PREEMPT Thu Jan 5 12:03:08 GMT 2023 **aarch64 GNU/Linux**; this is Raspberry Pi OS Lite 64-bit; I will replace the OS later once I deploy my boxes. I have more testing to do with owntone.
